tomoxlending/order_processor.go

```diff
@@ -861,62 +861,58 @@ func (l *Lending) GetCollateralPrices(header *types.Header, chain consensus.Chai
+	collateralPriceUpdatedFromContract := updatedBlock.Uint64()/chain.Config().Posv.Epoch == header.Number.Uint64()/chain.Config().Posv.Epoch
 
-	lendingTOMOBasePriceFromContract, lendingTOMOUpdatedBlock := lendingstate.GetCollateralPrice(statedb, lendingToken, common.HexToAddress(common.TomoNativeAddress))
-	lendingTOMOBasePriceUpdatedFromContract := lendingTOMOUpdatedBlock.Uint64()/chain.Config().Posv.Epoch == header.Number.Uint64()/chain.Config().Posv.Epoch
-
-	var lendTokenTOMOPrice *big.Int
-	var err error
-	if lendingToken == common.HexToAddress(common.TomoNativeAddress) {
-		lendTokenTOMOPrice = common.BasePrice
-	} else if lendingTOMOBasePriceUpdatedFromContract {
-		// getting lendToken price from contract first
-		// otherwise, getting from tomox lendToken/TOMO
-		log.Debug("Getting lendTokenTOMOPrice from contract", "lendTokenTOMOPrice", lendTokenTOMOPrice)
-		lendTokenTOMOPrice = lendingTOMOBasePriceFromContract
-	} else {
-		lendTokenTOMOPrice, err = l.GetMediumTradePriceBeforeEpoch(chain, statedb, tradingStateDb, lendingToken, common.HexToAddress(common.TomoNativeAddress))
-		if err != nil {
-			return lendTokenTOMOPrice, collateralPriceFromContract, err
-		}
-		log.Debug("Getting lendTokenTOMOPrice from tomox", "lendTokenTOMOPrice", lendTokenTOMOPrice, "err", err)
+	lendTokenTOMOPrice, err := l.GetTOMOBasePrices(header, chain, statedb, tradingStateDb, lendingToken)
+	if err != nil {
+		return nil, nil, err
 	}
 	if collateralPriceUpdatedFromContract {
-		log.Debug("Getting collateralPrice from contract", "collateralPrice", collateralPriceFromContract)
+		log.Debug("Getting collateral/lending token price from contract", "price", collateralPriceFromContract)
 		return lendTokenTOMOPrice, collateralPriceFromContract, nil
 	}
+	lendingTokenDecimal, err := l.tomox.GetTokenDecimal(chain, statedb, common.Address{}, lendingToken)
+	log.Debug("GetTokenDecimal", "lendingToken", lendingToken, "err", err)
+	if err != nil || lendingTokenDecimal == nil || lendingTokenDecimal.Sign() == 0 {
+		return nil, nil, err
+	}
+	collateralTokenDecimal, err := l.tomox.GetTokenDecimal(chain, statedb, common.Address{}, collateralToken)
+	log.Debug("GetTokenDecimal", "collateralToken", collateralToken, "err", err)
+	if err != nil || collateralTokenDecimal == nil || collateralTokenDecimal.Sign() == 0 {
+		return nil, nil, err
+	}
+	var collateralPrice *big.Int
+	inverseCollateralPriceFromContract, updatedBlock := lendingstate.GetCollateralPrice(statedb, lendingToken, collateralToken)
+	inverseCollateralPriceUpdatedFromContract := updatedBlock.Uint64()/chain.Config().Posv.Epoch == header.Number.Uint64()/chain.Config().Posv.Epoch
+	if inverseCollateralPriceUpdatedFromContract {
+		log.Debug("Getting lending/collateral token price from contract", "price", inverseCollateralPriceFromContract)
+		collateralPrice = new(big.Int).Mul(inverseCollateralPriceFromContract, collateralTokenDecimal)
+		collateralPrice = new(big.Int).Div(collateralPrice, lendingTokenDecimal)
```

Comment:
check this formula
I think it should be
```
collateralPrice = lendTokenDecimal * collateralTokenDecimal / inverseCollateralPriceFromContract
```

Reply:
updated